omp/matrix/fft_kernels.cpp

```diff
@@ -0,0 +1,334 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/fft_kernels.hpp"
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The FFT matrix format namespace.
+ * @ref Fft
+ * @ingroup fft
+ */
+namespace fft {
+
+
+template <typename InValueType, typename OutValueType>
+void bfly(const matrix::Dense<InValueType> *b, matrix::Dense<OutValueType> *x,
+          int64 lo, int64 hi, OutValueType root)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        auto lo_val = b->at(lo, rhs);
+        auto hi_val = b->at(hi, rhs);
+        x->at(lo, rhs) = lo_val + hi_val;
+        x->at(hi, rhs) = (lo_val - hi_val) * root;
+    }
+}
+
+
+template <typename ValueType>
+void bfly(matrix::Dense<ValueType> *x, int64 lo, int64 hi, ValueType root)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        auto lo_val = x->at(lo, rhs);
+        auto hi_val = x->at(hi, rhs);
+        x->at(lo, rhs) = lo_val + hi_val;
+        x->at(hi, rhs) = (lo_val - hi_val) * root;
```

Comment:
why this is not `lo_val +/- hi_val * root`?

Reply:
If you write `\sum_i=0^2n-1 x_i \omega_N^ik = \sum_i=0^n-1 (x_i + \omega_2n^nk x_{i+n}) \omega_2n^ik`, for even k=2k' you get `\omega_2n^2nk' = 1`, i.e. (x_i + x_i+n)\omega_n^ik, for odd `k` you get -1 from the inner omega and `root = \omega_2n^i` from the outer omega.

Reply:
I see